8304996: Add missing HandleMarks

dholmes-ora opened the pull request:
The review for [JDK-8304147](https://bugs.openjdk.org/browse/JDK-8304147) pointed out that the top-level HandleMark in dump_for_exit (added to replace the previous coverage from a HandleMarkCleaner in JVM_ENTRY) was not in the right place as no Handles were being used there. Removing that HM and fixing up the ensuing failures led to a set of fixes where HM's were missing at the place of Handle usage.

Testing: tiers 1-4

Thanks.

<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304996](https://bugs.openjdk.org/browse/JDK-8304996): Add missing HandleMarks


### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [0ffe0f20](https://git.openjdk.org/jdk/pull/13215/files/0ffe0f2066337e71789fe2c23138dcac0045081b)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13215/head:pull/13215` \
`$ git checkout pull/13215`

Update a local copy of the PR: \
`$ git checkout pull/13215` \
`$ git pull https://git.openjdk.org/jdk.git pull/13215/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13215`

View PR using the GUI difftool: \
`$ git pr show -t 13215`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13215.diff">https://git.openjdk.org/jdk/pull/13215.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13215#issuecomment-1487658795)

dholmes-ora commented:
/label add hotspot-runtime
/label remove hotspot

coleenp reviewed (approved):
Looks good! thank you for fixing these.

dholmes-ora commented:
Thanks @coleenp !

shipilev commented on src/hotspot/share/oops/klassVtable.cpp:
```diff
@@ -1174,6 +1175,7 @@ void klassItable::initialize_itable(GrowableArray<Method*>* supers) {
 void klassItable::check_constraints(GrowableArray<Method*>* supers, TRAPS) {
 
   assert(_size_method_table == supers->length(), "wrong size");
+  HandleMark hm(THREAD);
```
Question: Would this make the handles area grow up to `_size_method_table` in the worst case? I thought we want to have `HandleMark`-s closer to actual handle creation/last-use to keep the footprint at bay.

dholmes-ora replied:
I was optimising for runtime performance rather than footprint, but you are right this could grow in a way we don't want. So I reverted this change and made a similar change to `klassVtable::check_constraints`.

Thanks for looking at this @shipilev .

shipilev reviewed (commented):
I think the issue synopsis should be capitalized: "Add missing HandleMarks". Also, does it cover the cases that were not covered by `dump_at_exit` `HandleMark` before? If not, then it is not adding "missing" HMs.

dholmes-ora commented:
> I think the issue synopsis should be capitalized: "Add missing HandleMarks". Also, does it cover the cases that were not covered by `dump_at_exit` `HandleMark` before? If not, then it is not adding "missing" HMs.

Fixed capitalization in HandleMark.

Not sure what your issue is with "missing" - the HM's were not present where the Handles were used and hence were missing. Some cases were directly detected by removing the old HM in `dump_for_exit` but I also checked those files for other cases and added them.

iklam commented:
What are the rules for using HandleMarks? In the HotSpot code, there are plenty of cases where we create a HandleMark without creating any Handles in the immediately following code. There are also plenty of cases where Handles are created in functions that do not have any HandleMarks.

dholmes-ora commented:
> What are the rules for using HandleMarks? In the HotSpot code, there are plenty of cases where we create a HandleMark without creating any Handles in the immediately following code. There are also plenty of cases where Handles are created in functions that do not have any HandleMarks.

@coleenp has a lot to say on this but basically HM's should appear along-side the code that creates the Handles. Unless you return a Handle from a function, there should generally be a HM in the function. There is a lot of code that doesn't follow this. Partly I think this is because we have some high-on-the-stack HMs or HandleMarkCleaners that hide the creation of Handles without a local HandleMark. Conversely if new code uses existing code that doesn't have a HM and gets an error, then the typical response is to put a HM in the new code - especially because that only affects the new change, whereas putting a HM lower-down could be seen as disruptive etc.

shipilev commented:
> Not sure what your issue is with "missing" - the HM's were not present where the Handles were used and hence were missing. Some cases were directly detected by removing the old HM in `dump_for_exit` but I also checked those files for other cases and added them.

When I read a synopsis "Add missing HMs", that means to me there are code paths that allocate Handles without HMs -- meaning there is a memory leak. If we just dropped an "umbrella" HM in `dump_for_exit` and only added the HMs that were needed after that removal, that just rebalances the ways we cut and slice the HMs, and that does not affect memory leaks. This distinction is important to gauge how serious the issue is, and whether it should be backported, for example. This is why I wanted to understand the meaning of "missing" here :)

dholmes-ora commented:
Thanks @shipilev !

dholmes-ora commented:
Any other comments? @iklam or @coleenp ?

iklam reviewed (approved):
LGTM

dholmes-ora commented:
Thanks @iklam !

/integrate
